SwiftCompilerSources/Sources/Optimizer/Utilities/WalkUtils.swift

```diff
@@ -0,0 +1,422 @@
+      }
+      return false
+    }
+  }
+}
+
+protocol UseDefWalker {
+  mutating
+  func shouldRecomputeUp(value def: Value, path: SmallProjectionPath) -> SmallProjectionPath?
+  
+  mutating
+  func visitDef(ofValue value: Value, path: SmallProjectionPath,
+                isRoot: Bool) -> WalkResult
+  
+  mutating
+  func visitDef(ofAddress: Value, path: SmallProjectionPath,
+                isRoot: Bool) -> WalkResult
+}
+
+extension UseDefWalker {
+  mutating
+  func visitDef(ofValue value: Value, path: SmallProjectionPath,
+                isRoot: Bool)  -> WalkResult {
+    return .continueWalk
+  }
+  
+  mutating
+  func visitDef(ofAddress: Value, path: SmallProjectionPath, isRoot: Bool) -> WalkResult {
+    return .continueWalk
+  }
+  
+  /// - Returns: true if the walk was aborted
+  // Note: keep it symmetric to `DefUseWalker.walkDown(value:`
+  mutating
+  func walkUp(value def: Value, path: SmallProjectionPath) -> Bool {
+    var current: (value: Value, path: SmallProjectionPath) = (def, path)
+    
+    while true {
+      
+      var maybeNext: (Value, SmallProjectionPath)?
+      
+      let (value, path) = current
+      switch value {
+      // MARK: (trivially Non-Address to Non-Address) Constructions
+      case let str as StructInst:
+        if let (index, newPath) = path.pop(kind: .structField) {
+          maybeNext = (str.operands[index].value, newPath)
+        } else if path.topMatchesAnyValueField {
+          return visitAndWalkUp(value: value, path: path, operands: str.operands, newPath: path)
+        }
+      case let t as TupleInst:
+        if let (index, newPath) = path.pop(kind: .tupleField) {
+          maybeNext = (t.operands[index].value, newPath)
+        } else if path.topMatchesAnyValueField {
+          return visitAndWalkUp(value: value, path: path, operands: t.operands, newPath: path)
+        }
+      case let e as EnumInst:
+        if let newPath = path.popIfMatches(.enumCase, index: e.caseIndex),
+           let operand = e.operand {
+          maybeNext = (operand, newPath)
+        }
+      // MARK: Non-Address to Non-Address Projections
+      case let se as StructExtractInst:
+        maybeNext = (se.operand, path.push(.structField, index: se.fieldIndex))
+      case let te as TupleExtractInst:
+        maybeNext = (te.operand, path.push(.tupleField, index: te.fieldIndex))
+      case let ued as UncheckedEnumDataInst:
+        maybeNext = (ued.operand, path.push(.enumCase, index: ued.caseIndex))
+      // MARK: (trivially Non-Address to Non-Address) Multiresult Projections
+      case let mvr as MultipleValueInstructionResult
+        where mvr.instruction is DestructureStructInst:
+        let ds = mvr.instruction as! DestructureStructInst
+        maybeNext = (ds.operand, path.push(.structField, index: mvr.index))
+      case let mvr as MultipleValueInstructionResult
+        where mvr.instruction is DestructureTupleInst:
+        let dt = mvr.instruction as! DestructureTupleInst
```

Comment:
I really prefer to a `let ... as ...` instead of a `is` + `as!`.
That's probably only possible to have a top-level case for MultipleValueInstructionResult and then a sub-switch for the individual instructions (like in EscapeInfo now).